### PR TITLE
use MaybeUninit instead of mem::uninitialized for Windows Mutex

### PR DESCRIPTION
**Pull request by RalfJung**
I hope this builds, I do not have a Windows machine to test...

### CODE REVIEW

File: src/libstd/sys/windows/mutex.rs [hunk 1]
```diff
@@ -157,34 +157,34 @@ fn kind() -> Kind {
...
+pub struct ReentrantMutex { inner: MaybeUninit<UnsafeCell<c::CRITICAL_SECTION>> }
 
 unsafe impl Send for ReentrantMutex {}
 unsafe impl Sync for ReentrantMutex {}
 
 impl ReentrantMutex {
-    pub unsafe fn uninitialized() -> ReentrantMutex {
-        mem::uninitialized()
+    pub fn uninitialized() -> ReentrantMutex {
+        MaybeUninit::uninitialized()
```
Comment (abonander) [inline src/libstd/sys/windows/mutex.rs]: Looks like you forgot to wrap this in `ReentrantMutex`.

**Comment by Mark-Simulacrum** [comment]
FWIW, you should be able to run the mingw-check builder locally (`src/ci/docker/run.sh mingw-check` I believe) if you have Docker.

Code-wise this seems fine but someone else should probably review as well as I'm not up to date on `MaybeUninit` myself -- not sure who to assign though personally.

File: src/libstd/sys/windows/mutex.rs (hunk 1 above)
Comment (RalfJung) [reply src/libstd/sys/windows/mutex.rs]: I forgot indeed, thanks!

File: src/libstd/sys/windows/mutex.rs [hunk 2]
```diff
@@ -157,34 +157,37 @@ fn kind() -> Kind {
     return ret;
 }
 
-pub struct ReentrantMutex { inner: UnsafeCell<c::CRITICAL_SECTION> }
+pub struct ReentrantMutex { inner: MaybeUninit<UnsafeCell<c::CRITICAL_SECTION>> }
```
Comment (nagisa) [inline src/libstd/sys/windows/mutex.rs]: Would it not make more sense to keep `MaybeUninit` *within* `UsafeCell`?

Comment (RalfJung) [reply src/libstd/sys/windows/mutex.rs]: Good point, fixed.

**Comment by RalfJung** [comment]
> Code-wise this seems fine but someone else should probably review as well as I'm not up to date on MaybeUninit myself -- not sure who to assign though personally.

@SimonSapin could you have a look?

File: src/libstd/sys/windows/mutex.rs [hunk 3]
```diff
@@ -157,34 +157,42 @@ fn kind() -> Kind {
...
+        // `init` must have been called, so this is now initialized and
+        // we can call `get_mut`.
+        c::LeaveCriticalSection((&mut *self.inner.get()).get_mut());
     }
 
     pub unsafe fn destroy(&self) {
-        c::DeleteCriticalSection(self.inner.get());
+        // `init` must have been called, so this is now initialized and
+        // we can call `get_mut`.
+        c::DeleteCriticalSection((&mut *self.inner.get()).get_mut());
```
Comment (SimonSapin) [inline src/libstd/sys/windows/mutex.rs]: These `get_mut` calls could be `as_mut_ptr`, right? This avoids the unresolved question of uninit `&mut T`.

File: src/libcore/fmt/float.rs [hunk 4]
```diff
@@ -22,6 +22,9 @@ fn float_to_decimal_common_exact<T>(fmt: &mut Formatter, num: &T,
     unsafe {
         let mut buf = MaybeUninit::<[u8; 1024]>::uninitialized(); // enough for f32 and f64
         let mut parts = MaybeUninit::<[flt2dec::Part; 4]>::uninitialized();
+        // FIXME: Technically, this is calling `get_mut` on an uninitialized
```
Comment (SimonSapin) [inline src/libcore/fmt/float.rs]: Maybe file an issue about this specifically, and mention that issue number in code comments for all known occurences? That way when the question is resolved we can grep for the issue number to find places to fix (or to remove outdated FIXME comments).

File: src/libstd/sys/windows/mutex.rs (hunk 3 above)
Comment (RalfJung) [reply src/libstd/sys/windows/mutex.rs]: It is incorrect to call any of these methods without previously having initialized the mutex. So I don't think the question even comes up.

But you are also right that there is no reason to use `&mut` at all here, so I will change this to `*mut`.

File: src/libcore/fmt/float.rs (hunk 4 above)
Comment (RalfJung) [reply src/libcore/fmt/float.rs]: Wouldn't this usually be part of the discussion in the [tracking issue](https://github.com/rust-lang/rust/issues/53491)?

Comment (RalfJung) [reply src/libcore/fmt/float.rs]: I agree. Is like this okay? This covers all `get_mut` uses that got added as part of my PR series. I grepped the codebase for other uses of `MaybeUninit` and found none.

**Comment by SimonSapin** [comment]
Looks good, thanks!

@bors r+

**Pull request merged**